File: Algorithm.Framework/Alphas/PairsTradingAlphaModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using QuantConnect.Data;
using QuantConnect.Data.UniverseSelection;
using QuantConnect.Indicators;
using QuantConnect.Securities;

namespace QuantConnect.Algorithm.Framework.Alphas
{
    /// <summary>
    /// This alpha model is designed to work against a single, predefined pair.
    /// This model generates alternating long ratio/short ratio insights emitted as a group
    /// </summary>
    public class PairsTradingAlphaModel : AlphaModel
    {
        private readonly TimeSpan _period;
        private readonly decimal _threshold;

        private readonly HashSet<Security> _securities;
        private readonly Dictionary<string, Pair> _pairs;

        /// <summary>
        /// Initializes a new instance of the <see cref="PairsTradingAlphaModel"/> class
        /// </summary>
        /// <param name="period">Period over which this insight is expected to come to fruition</param>
        /// <param name="threshold">The percent [0, 100] deviation of the ratio from the mean before emitting an insight</param>
        public PairsTradingAlphaModel(TimeSpan period, decimal threshold = 1m)
        {
            _period = period;
            _threshold = threshold;

            _securities = new HashSet<Security>();
            _pairs = new Dictionary<string, Pair>();

            Name = $"{nameof(PairsTradingAlphaModel)}({_period},{_threshold.Normalize()})";
        }

        /// <summary>
        /// Updates this alpha model with the latest data from the algorithm.
        /// This is called each time the algorithm receives data for subscribed securities
        /// </summary>
        /// <param name="algorithm">The algorithm instance</param>
        /// <param name="data">The new data available</param>
        /// <returns>The new insights generated</returns>
        public override IEnumerable<Insight> Update(QCAlgorithmFramework algorithm, Slice data)
        {
            var insights = new List<Insight>();

            foreach (var kvp in _pairs)
            {
                var pair = kvp.Value;

                if (pair.IsReady)
                {
                    insights.AddRange(pair.GetInsightGroup());
                }
            }

            return insights;
        }

        /// <summary>
        /// Event fired each time the we add/remove securities from the data feed
        /// </summary>
        /// <param name="algorithm">The algorithm instance that experienced the change in securities</param>
        /// <param name="changes">The security additions and removals from the algorithm</param>
        public override void OnSecuritiesChanged(QCAlgorithmFramework algorithm, SecurityChanges changes)
        {
            NotifiedSecurityChanges.UpdateCollection(_securities, changes);

            UpdatePairs(algorithm);

            // Remove pairs that has assets that were removed from the universe 
            foreach (var security in changes.RemovedSecurities)
            {
                var assetId = security.Symbol.ID.ToString();
                var keys = _pairs.Keys.Where(k => k.Contains(assetId));

                foreach (var key in keys)
                {
                    _pairs.Remove(key);
                }
            }
        }

        private void UpdatePairs(QCAlgorithm algorithm)
        {
            var securities = _securities.ToArray();

            for (var i = 0; i < securities.Length; i++)
            {
                var assetI = securities[i].Symbol;

                for (var j = i + 1; j < securities.Length; j++)
                {
                    var assetJ = securities[j].Symbol;

                    var pairName = Pair.GetPairName(assetI, assetJ);

                    if (_pairs.ContainsKey(pairName))
                    {
                        continue;
                    }

                    if (!HasPassedTest(algorithm, assetI, assetJ))
                    {
                        continue;
                    }

                    var pair = new Pair(algorithm, assetI, assetJ, _period, _threshold);
                    _pairs.Add(pairName, pair);
                }
            }
        }

        /// <summary>
        /// Check whether the assets pass a pairs trading test
        /// </summary>
        /// <param name="algorithm">The algorithm instance that experienced the change in securities</param>
        /// <param name="asset1">The first asset's symbol in the pair</param>
        /// <param name="asset2">The second asset's symbol in the pair</param>
        /// <returns>True if the statistical test for the pair is successful</returns>
        public virtual bool HasPassedTest(QCAlgorithm algorithm, Symbol asset1, Symbol asset2) => true;

        internal class Pair
        {
            private enum State
            {
                ShortRatio,
                FlatRatio,
                LongRatio
            };

            /// <summary>
            /// Creates a name for the pair
            /// </summary>
            /// <param name="asset1">The first asset's symbol in the pair</param>
            /// <param name="asset2">The second asset's symbol in the pair</param>
            /// <returns>String: name of the pair</returns>
            public static string GetPairName(Symbol asset1, Symbol asset2)
            {
                var asset1Id = asset1.ID.ToString();
                var asset2Id = asset2.ID.ToString();

                return asset1Id.CompareTo(asset2Id) < 0
                    ? $"{asset1Id}/{asset2Id}"
                    : $"{asset2Id}/{asset1Id}";
            }

            private State _state = State.FlatRatio;

            private readonly Symbol _asset1;
            private readonly Symbol _asset2;

            private readonly IndicatorBase<IndicatorDataPoint> _asset1Price;
            private readonly IndicatorBase<IndicatorDataPoint> _asset2Price;
            private readonly IndicatorBase<IndicatorDataPoint> _ratio;
            private readonly IndicatorBase<IndicatorDataPoint> _mean;
            private readonly IndicatorBase<IndicatorDataPoint> _upperThreshold;
            private readonly IndicatorBase<IndicatorDataPoint> _lowerThreshold;

            private readonly TimeSpan _period;

            public bool IsReady => _mean.IsReady;

            /// <summary>
            /// Create a new pair
            /// </summary>
            /// <param name="algorithm">The algorithm instance that experienced the change in securities</param>
            /// <param name="asset1">The first asset's symbol in the pair</param>
            /// <param name="asset2">The second asset's symbol in the pair</param>
            /// <param name="period">Period over which this insight is expected to come to fruition</param>
            /// <param name="threshold">The percent [0, 100] deviation of the ratio from the mean before emitting an insight</param>
            public Pair(QCAlgorithm algorithm, Symbol asset1, Symbol asset2, TimeSpan period, decimal threshold)
            {
                _asset1 = asset1;
                _asset2 = asset2;

                _asset1Price = algorithm.Identity(asset1);
                _asset2Price = algorithm.Identity(asset2);

                _ratio = _asset1Price.Over(_asset2Price);
                _mean = new ExponentialMovingAverage(500).Of(_ratio);

                var upper = new ConstantIndicator<IndicatorDataPoint>("ct", 1 + threshold / 100m);
                _upperThreshold = _mean.Times(upper, "UpperThreshold");

                var lower = new ConstantIndicator<IndicatorDataPoint>("ct", 1 - threshold / 100m);
                _lowerThreshold = _mean.Times(lower, "LowerThreshold");

                _period = period;
            }

            /// <summary>
            /// Gets the insights group for the pair
            /// </summary>
            /// <returns>Insights grouped by an unique group id</returns>
            public IEnumerable<Insight> GetInsightGroup()
            {
                // don't re-emit the same direction
                if (_state != State.LongRatio && _ratio > _upperThreshold)
                {
                    _state = State.LongRatio;

                    // asset1/asset2 is more than 2 std away from mean, short asset1, long asset2
                    var shortAsset1 = Insight.Price(_asset1, _period, InsightDirection.Down);
                    var longAsset2 = Insight.Price(_asset2, _period, InsightDirection.Up);

                    // creates a group id and set the GroupId property on each insight object
                    return Insight.Group(shortAsset1, longAsset2);
                }

                // don't re-emit the same direction
                if (_state != State.ShortRatio && _ratio < _lowerThreshold)
                {
                    _state = State.ShortRatio;

                    // asset1/asset2 is less than 2 std away from mean, long asset1, short asset2
                    var longAsset1 = Insight.Price(_asset1, _period, InsightDirection.Up);
                    var shortAsset2 = Insight.Price(_asset2, _period, InsightDirection.Down);

                    // creates a group id and set the GroupId property on each insight object
                    return Insight.Group(longAsset1, shortAsset2);
                }

                return Enumerable.Empty<Insight>();
            }
        }
    }
}

File: Algorithm.Framework/Alphas/PairsTradingAlphaModel.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from clr import AddReference
AddReference("QuantConnect.Common")
AddReference("QuantConnect.Algorithm.Framework")
AddReference("QuantConnect.Indicators")

from QuantConnect import *
from QuantConnect.Indicators import *
from QuantConnect.Algorithm.Framework.Alphas import *
from datetime import timedelta
from enum import Enum

class PairsTradingAlphaModel(AlphaModel):
    '''This alpha model is designed to work against a single, predefined pair.
    This model generates alternating long ratio/short ratio insights emitted as a group'''

    class State(Enum):
        ShortRatio = -1
        FlatRatio = 0
        LongRatio = 1

    def __init__(self, asset1, asset2, threshold = 1):
        ''' Initializes a new instance of the PairsTradingAlphaModel class
        Args:
            asset1: The first asset's symbol in the pair
            asset2: The second asset's symbol in the pair
            threshold: The percent [0, 100] deviation of the ratio from the mean before emitting an insight'''
        self.asset1 = asset1
        self.asset2 = asset2
        self.threshold = threshold
        self.state = self.State.FlatRatio
        self.asset1Price = None
        self.asset2Price = None
        self.ratio = None
        self.mean = None
        self.upperThreshold = None
        self.lowerThreshold = None

        self.Name = '{}({},{},{})'.format(self.__class__.__name__, asset1, asset2, Extensions.Normalize(threshold))


    def Update(self, algorithm, data):
        ''' Updates this alpha model with the latest data from the algorithm.
        This is called each time the algorithm receives data for subscribed securities
        Args:
            algorithm: The algorithm instance
            data: The new data available
        Returns:
            The new insights generated'''
        if self.mean is None or not self.mean.IsReady:
            return []

        # don't re-emit the same direction
        if self.state is not self.State.LongRatio and self.ratio > self.upperThreshold:
            self.state = self.State.LongRatio

            # asset1/asset2 is more than 2 std away from mean, short asset1, long asset2
            shortAsset1 = Insight.Price(self.asset1, timedelta(minutes = 15), InsightDirection.Down)
            longAsset2 = Insight.Price(self.asset2, timedelta(minutes = 15), InsightDirection.Up)

            # creates a group id and set the GroupId property on each insight object
            return Insight.Group(shortAsset1, longAsset2)

        # don't re-emit the same direction
        if self.state is not self.State.ShortRatio and self.ratio < self.lowerThreshold:
            self.state = self.State.ShortRatio

            # asset1/asset2 is less than 2 std away from mean, long asset1, short asset2
            longAsset1 = Insight.Price(self.asset1, timedelta(minutes = 15), InsightDirection.Up)
            shortAsset2 = Insight.Price(self.asset2, timedelta(minutes = 15), InsightDirection.Down)

            # creates a group id and set the GroupId property on each insight object
            return Insight.Group(longAsset1, shortAsset2)

        return []

    def OnSecuritiesChanged(self, algorithm, changes):
        '''Event fired each time the we add/remove securities from the data feed.
        Args:
            algorithm: The algorithm instance that experienced the change in securities
            changes: The security additions and removals from the algorithm'''
        for added in changes.AddedSecurities:
            # this model is limitted to looking at a single pair of assets
            if added.Symbol != self.asset1 and added.Symbol != self.asset2:
                continue

            if added.Symbol == self.asset1:
                self.asset1Price = algorithm.Identity(added.Symbol)
            else:
                self.asset2Price = algorithm.Identity(added.Symbol)

        if self.ratio is None:
            # initialize indicators dependent on both assets
            if self.asset1Price is not None and self.asset2Price is not None:
                self.ratio = IndicatorExtensions.Over(self.asset1Price, self.asset2Price)
                self.mean = IndicatorExtensions.Of(ExponentialMovingAverage(500), self.ratio)
                
                upper = ConstantIndicator[IndicatorDataPoint]("ct", 1 + self.threshold / 100)
                self.upperThreshold = IndicatorExtensions.Times(self.mean, upper)

                lower = ConstantIndicator[IndicatorDataPoint]("ct", 1 - self.threshold / 100)
                self.lowerThreshold = IndicatorExtensions.Times(self.mean, lower)